src/cache/mod.rs

```diff
@@ -372,8 +379,21 @@ impl<'a> Locatable<'a> for EffectInfo<'a> {
     }
 }
 
+pub fn cached_read(file_cache: &FileCache<'_>, path: &Path) -> Option<String> {
+    file_cache.get(path).cloned().or_else(|| {
+        let file = File::open(path).ok()?;
+        let mut reader = BufReader::new(file);
+        let mut contents = String::new();
+        reader.read_to_string(&mut contents).ok()?;
+
+        Some(contents)
+    })
+}
```

Comment:
We can avoid cloning a cached File using a `Cow`. Alternatively, maybe we should automatically add a non-cached file's contents to the cache automatically similar to `get_contents`?
```suggestion
pub fn cached_read<'c>(file_cache: &FileCache<'c>, path: &Path) -> Option<Cow<'c, str>> {
    file_cache.get(path).map_or_else(Cow::Borrowed, || {
        let file = File::open(path).ok()?;
        let mut reader = BufReader::new(file);
        let mut contents = String::new();
        reader.read_to_string(&mut contents).ok()?;
        Some(Cow::Owned(contents))
    })
}
```